File: examples/golang_examples/ping_pong.go
// Taken from Googler Sameer Ajmani's talk _Advanced Go Concurrency Patterns_
// available at https://talks.golang.org/2013/advconc.slide#6 What's nice about
// this example is its simplicity
package main

import (
	"fmt"
	"time"
)

type Ball struct{ hits int }

func player(name string, table chan *Ball) {
	for {
		ball := <-table
		ball.hits++
		fmt.Println(name, ball.hits)
		time.Sleep(100 * time.Millisecond)
		table <- ball
	}
}

func main() {
	table := make(chan *Ball)
	go player("ping", table)
	go player("pong", table)
	table <- new(Ball) // game on; toss the ball
	time.Sleep(1 * time.Second)
	<-table // game over; grab the ball
}

File: examples/golang_examples/ping_pong.lua
package.path = "../../?.lua;../?.lua;" .. package.path

local channel = require 'fibers.channel'
local fiber = require 'fibers.fiber'
local sleep = require 'fibers.sleep'

--- so simple to recreate Go's go
local function go(fn, args)
    fiber.spawn(function ()
        fn(unpack(args or {}))
    end)
end

local function player(name, board)
    while true do
        local ball = board:get()
        ball = ball + 1
        print(name, ball)
        sleep.sleep(.1)
        board:put(ball)
    end
end

local function main()
    local ball = 0
    local board = channel.new()
    go(player, {'ping', board})
    go(player, {'pong', board})
    board:put(ball) -- game on; toss the ball
    sleep.sleep(1)
    board:get() -- game over; grab the ball
    fiber.current_scheduler.done = true
end

fiber.spawn(main)
fiber.current_scheduler:main()
